### PR TITLE
Support for data validation

### PR DESCRIPTION
**Pull request by emilaxelsson**
I wasn't sure whether data validation should also be supported by the "Formatted" API, but I just implemented basic support for now.

### CODE REVIEW

**Comment by qrilka** [comment]
thanks @emilaxelsson I'll try to review it later today

File: src/Codec/Xlsx/Types/ConditionalFormatting.hs [hunk 1]
```diff
@@ -265,19 +267,24 @@ conditionData (EndsWith t)           = ("endsWith", M.fromList [ "text" .= t], [
...
-operatorExpressionData (OpEndsWith f)            = ("endsWith", [formulaNode f])
-operatorExpressionData (OpEqual f)               = ("equal", [formulaNode f])
-operatorExpressionData (OpGreaterThan f)         = ("greaterThan", [formulaNode f])
-operatorExpressionData (OpGreaterThanOrEqual f)  = ("greaterThanOrEqual", [formulaNode f])
-operatorExpressionData (OpLessThan f)            = ("lessThan", [formulaNode f])
-operatorExpressionData (OpLessThanOrEqual f)     = ("lessThanOrEqual", [formulaNode f])
-operatorExpressionData (OpNotBetween f1 f2)      = ("notBetween", [formulaNode f1, formulaNode f2])
-operatorExpressionData (OpNotContains f)         = ("notContains", [formulaNode f])
-operatorExpressionData (OpNotEqual f)            = ("notEqual", [formulaNode  f])
+operatorExpressionData oe = (op, map formulaNode $ catMaybes [f1,f2])
```
Comment (qrilka) [inline src/Codec/Xlsx/Types/ConditionalFormatting.hs]: what about `mapMaybe`?

File: src/Codec/Xlsx/Types/DataValidation.hs [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+import           Codec.Xlsx.Parser.Internal
+import           Codec.Xlsx.Types.Common
+import           Codec.Xlsx.Types.ConditionalFormatting
+import           Codec.Xlsx.Writer.Internal
+
+-- See 18.18.21 "ST_DataValidationType (Data Validation Type)" (p. 2440/2450)
+data ValidationType
+    = ValidationTypeNone
+    | ValidationTypeCustom     Formula
+    | ValidationTypeDate       OperatorExpression
```
Comment (qrilka) [inline src/Codec/Xlsx/Types/DataValidation.hs]: this isn't quite correct `ST_DataValidationOperator` != `ST_ConditionalFormattingOperator` unfortunately :(

File: src/Codec/Xlsx/Types/DataValidation.hs [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+            _dvErrorStyle       <- maybeAttribute "errorStyle"       c
+            _dvErrorTitle       <- maybeAttribute "errorTitle"       c
+            mop                 <- maybeAttribute "operator"         c
+            _dvPrompt           <- maybeAttribute "prompt"           c
+            _dvPromptTitle      <- maybeAttribute "promptTitle"      c
+            _dvShowDropDown     <- maybeAttribute "showDropDown"     c
+            _dvShowErrorMessage <- maybeAttribute "showErrorMessage" c
+            _dvShowInputMessage <- maybeAttribute "showInputMessage" c
+            _dvSqref            <- fromAttribute  "sqref"            c
+            mtype               <- maybeAttribute "type"             c
```
Comment (qrilka) [inline src/Codec/Xlsx/Types/DataValidation.hs]: This attribute has default value of `none` so this field doesn't need `Maybe` e.g.
Also see the defaults for `errorStyle`, `allowBlank`, `showDropDown`, `showInputMessage` and `showErrorMessage` in schema file `sml.rnc`

File: src/Codec/Xlsx/Types/DataValidation.hs [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+            _dvValidationType   <- readValidation mop mtype          c
+            return DataValidation{..}
+
+readValidation :: Maybe Text -> Maybe Text -> Cursor -> [Maybe ValidationType]
```
Comment (qrilka) [inline src/Codec/Xlsx/Types/DataValidation.hs]: this name is somewhat confusing, let's use something more explicit, e.g. `readValidationType` maybe?

File: src/Codec/Xlsx/Types/DataValidation.hs [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+readValidation _ (Just "list") cur = do
+    f  <- cur $/ element (n_ "formula1") >=> fromCursor
+    as <- readListFormula f
+    return $ Just $ ValidationTypeList as
+readValidation (Just op) (Just ty) cur = do
+    opExp <- readOpExpression2 op cur
+    Just <$> readValidationTypeOpExp ty opExp
+readValidation _ _ _ = []
+
+readListFormula :: Formula -> [[Text]]
```
Comment (qrilka) [inline src/Codec/Xlsx/Types/DataValidation.hs]: what's the point of wrapping text value into `Formula` just to unwrap it afterwards?

File: src/Codec/Xlsx/Types/DataValidation.hs [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+-------------------------------------------------------------------------------}
+
+instance FromAttrVal ErrorStyle where
+    fromAttrVal "information" = readSuccess ErrorStyleInformation
+    fromAttrVal "stop"        = readSuccess ErrorStyleStop
+    fromAttrVal "warning"     = readSuccess ErrorStyleWarning
+    fromAttrVal t             = invalidText "ErrorStyle" t
+
+instance FromCursor DataValidation where
+    fromCursor cur = cur $/ element (n_ "dataValidation") >=> fromCur
```
Comment (qrilka) [inline src/Codec/Xlsx/Types/DataValidation.hs]: that doesn't look correct - `fromCursor` is supposed to be invoked with cursor pointing to the node which should give you results, i.e. it should already point to `<dataValidation>`, i.e. you miss that level in `Parser`

**Comment by emilaxelsson** [comment]
Thanks for a quick review! I will try to fix these things today.

File: src/Codec/Xlsx/Types/DataValidation.hs (hunk 3 above)
Comment (emilaxelsson) [inline src/Codec/Xlsx/Types/DataValidation.hs]: OK, so should I remove `Maybe` from all optional fields that have a default value then?

File: src/Codec/Xlsx/Types/DataValidation.hs (hunk 5 above)
Comment (qrilka) [inline src/Codec/Xlsx/Types/DataValidation.hs]: For `Text` there is https://www.stackage.org/haddock/lts-7.5/xml-conduit-1.3.5/Text-XML-Cursor.html#v:content but probably this point isn't that important

File: src/Codec/Xlsx/Types/DataValidation.hs (hunk 3 above)
Comment (qrilka) [inline src/Codec/Xlsx/Types/DataValidation.hs]: Yes, remove `Maybe` and use `fromAttributeDef`

File: src/Codec/Xlsx/Parser.hs [hunk 7]
```diff
@@ -148,14 +149,20 @@ extractSheet ar sst contentTypes wf = do
 
       condFormtattings = M.fromList . map unCfPair  $ cur $/ element (n_ "conditionalFormatting") >=> fromCursor
 
+      validations = M.fromList . map unDvPair $
+          ( ($/ element (n_ "dataValidations")) >=>
```
Comment (qrilka) [inline src/Codec/Xlsx/Parser.hs]: that's a bit a bit unexpected way to use `xml-conduit`, why not `cur $/ element (n_ "dataValidations") &/ element (n_ "dataValidation") >=> fromCursor` ?

File: src/Codec/Xlsx/Writer.hs [hunk 8]
```diff
@@ -120,12 +121,14 @@ singleSheetFiles n cells ws = runST $ do
             , Just . elementListSimple "sheetData" $ sheetDataXml cells (ws ^. wsRowPropertiesMap)
             ] ++
             map (Just . toElement "conditionalFormatting") cfPairs ++
-            [ nonEmptyElListSimple "mergeCells" . map mergeE1 $ ws ^. wsMerges
+            [ Just $ countedElementList "dataValidations" $ map (toElement "dataValidation") dvPairs
```
Comment (qrilka) [inline src/Codec/Xlsx/Writer.hs]: please see sml.rnc, specifically `sml_CT_Worksheet` - `dataValidations` are expected to be after `mergeCells` and before `pageSetup`

Comment (emilaxelsson) [inline src/Codec/Xlsx/Writer.hs]: Ah, I didn't realize that. But then it seems like `conditionalFormatting` is also incorrectly placed, no? It should come right before `dataValidations`.

**Pull request closed without merging**